pg_search/src/api/operator.rs

```diff
@@ -15,74 +15,237 @@
+    }
+}
+
+fn estimate_selectivity(heaprelid: pg_sys::Oid, search_config: &SearchConfig) -> Option<f64> {
+    let reltuples = unsafe { PgRelation::open(heaprelid).reltuples().unwrap_or(1.0) as f64 };
+    if !reltuples.is_normal() || reltuples.is_sign_negative() {
+        // we can't estimate against a non-normal or negative estimate of heap tuples
+        return None;
+    }
+
+    let directory = WriterDirectory::from_index_oid(search_config.index_oid);
+    let search_index = SearchIndex::from_cache(&directory, &search_config.uuid)
+        .unwrap_or_else(|err| panic!("error loading index from directory: {err}"));
+    let writer_client = WriterGlobal::client();
+    let state = search_index
+        .search_state(&writer_client, search_config, false)
+        .expect("SearchState creation should not fail");
+    let estimate = state.estimate_docs().unwrap_or(1) as f64;
+
+    let mut selectivity = estimate / reltuples;
+    if selectivity > 1.0 {
+        selectivity = 1.0;
+    }
+
+    Some(selectivity)
+}
+
+unsafe fn make_search_config_opexpr_node(
+    srs: *mut SupportRequestSimplify,
+    input_args: &mut PgList<Node>,
+    var: *mut pg_sys::Var,
+    query: SearchQueryInput,
+) -> ReturnedNodePointer {
+    // we're about to fabricate a new pg_sys::OpExpr node to return
+    // that represents the `@@@(anyelement, jsonb)` operator
+    let mut newopexpr = pg_sys::OpExpr {
+        xpr: pg_sys::Expr {
+            type_: pg_sys::NodeTag::T_OpExpr,
+        },
+        opno: anyelement_jsonb_opoid(),
+        opfuncid: anyelement_jsonb_procoid(),
+        opresulttype: pg_sys::BOOLOID,
+        opretset: false,
+        opcollid: pg_sys::DEFAULT_COLLATION_OID,
+        inputcollid: pg_sys::DEFAULT_COLLATION_OID,
+        args: std::ptr::null_mut(),
+        location: (*(*srs).fcall).location,
     };
 
-    hash_set.contains(&key_field_value)
+    // we need to use what should be the only `USING bm25` index on the table
+    let rte = planner_rt_fetch((*var).varno as pg_sys::Index, (*srs).root);
+    let heaprel = PgRelation::open((*rte).relid);
+    let indexrel = locate_bm25_index((*rte).relid).unwrap_or_else(|| {
+        panic!(
+            "relation `{}.{}` must have a `USING bm25` index",
+            heaprel.namespace(),
+            heaprel.name()
+        )
+    });
+
+    let heaptupdesc = heaprel.tuple_desc();
+    let varatt = heaptupdesc
+        .get(
+            ((*var).varattno - 1)
+                .try_into()
+                .expect("varattno should be within the bounds of usize"),
+        )
+        .expect("varattno should exist in the heap");
+
+    // fabricate a `SearchConfig` from the above relation and query string
+    // and get it serialized into a JSONB Datum
+    let search_config = SearchConfig::from((query, indexrel));
+    if search_config.key_field != varatt.name() {
+        panic!("left-hand side of the @@@ operator should be the `key_field` of the only `USING bm25` index");
+    }
```

Comment:
I think the search_config is deserialized from disk, so the value of `search_config.key_field` will probably not change if the user renames the column.


I don't think this is the only place where this is an issue. The index itself doesn't have a way to keep track of the column name being changed, so there's more than one thing that will be broken. I just wanted to bring it to your attention.

Reply:
hmm.  I didn't consider what might happen if the user renames the column.  I'll consider what to do here.

I think the answer is to make sure the lhs's `pg_sys::Var` node matches the first column of the (only) `USING bm25` index.  Then we'd be resilient to column renames and such.

Reply:
Okay, the code now matches the lhs of the @@@ operator against the first column in the only bm25 index, by their attribute numbers.  No need to worry about names at all.

Thanks.  Good catch.